lib/web_ui/flutter_js/src/entrypoint_loader.js

```diff
@@ -0,0 +1,198 @@
+   * object when it's done.
+   *
+   * @param {*} options
+   * @returns {Promise | undefined} that will eventually resolve with an
+   * EngineInitializer, or will be rejected with the error caused by the loader.
+   * Returns undefined when an `onEntrypointLoaded` callback is supplied in `options`.
+   */
+  async loadEntrypoint(options) {
+    const { entrypointUrl = `${baseUri}main.dart.js`, onEntrypointLoaded, nonce } =
+      options || {};
+    return this._loadJSEntrypoint(entrypointUrl, onEntrypointLoaded, nonce);
+  }
+
+  /**
+   * Loads the entry point for a flutter application.
+   * @param {import("./types").ApplicationBuild} build
+   *   Information about the specific build that is to be loaded
+   * @param {*} deps
+   *   External dependencies that may be needed to load the app.
+   * @param {import("./types").FlutterConfiguration} config
+   *   The application configuration. If no callback is specified, this will be
+   *   passed along to engine when initializing it.
+   * @param {string} nonce
+   *   A nonce to apply to the main application script tag, if necessary.
+   * @param {import("./types").OnEntrypointLoadedCallback?} onEntrypointLoaded
+   *   An optional callback to invoke when the entrypoint is loaded. If no
+   *   callback is supplied, the engine initializer and app runner will be
+   *   automatically invoked on load, passing along the supplied flutter
+   *   configuration.
+   */
+  async load(build, deps, config, nonce, onEntrypointLoaded) {
+    onEntrypointLoaded ??= (engineInitializer) => {
+      engineInitializer.initializeEngine(config).then((appRunner) => appRunner.runApp())
+    };
+    if (build.compileTarget == "dart2wasm") {
+      return this._loadWasmEntrypoint(build, deps, onEntrypointLoaded);
+    } else {
+      const mainPath = build.mainJsPath ?? "main.dart.js";
+      const entrypointUrl = `${baseUri}${mainPath}`;
+      return this._loadJSEntrypoint(entrypointUrl, onEntrypointLoaded, nonce);
+    }
+  }
+
+  /**
+   * Resolves the promise created by loadEntrypoint, and calls the `onEntrypointLoaded`
+   * function supplied by the user (if needed).
+   *
+   * Called by Flutter through `_flutter.loader.didCreateEngineInitializer` method,
+   * which is bound to the correct instance of the FlutterEntrypointLoader by
+   * the FlutterLoader object.
+   *
+   * @param {Function} engineInitializer @see https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/js_interop/js_loader.dart#L42
+   */
+  didCreateEngineInitializer(engineInitializer) {
+    if (typeof this._didCreateEngineInitializerResolve === "function") {
+      this._didCreateEngineInitializerResolve(engineInitializer);
+      // Remove the resolver after the first time, so Flutter Web can hot restart.
+      this._didCreateEngineInitializerResolve = null;
+      // Make the engine revert to "auto" initialization on hot restart.
+      delete _flutter.loader.didCreateEngineInitializer;
+    }
```

Comment:
My take on this is that I don't actually want to change the behavior of `loadEntrypoint` at all with this change, if possible. The reason is that I think the next step for this API should be removal after deprecation. I don't really want to change the behavior of the API and then remove it. I think it would be better to make users migrate a single time and make a clean break, rather than subtly change the behavior of the deprecated API. If users are still using the promise-based method on `loadEntrypoint`, they will be gently nudged to move over to the `load` API (where they won't rely on promise-based behavior) and then maybe one more release later we'll forcibly remove it. At that point we can get rid of this weird code path. But for now I don't think we should mess with the existing behavior of the deprecated API.

Reply:
OK, let's keep for a while longer then!